webhooks/virtualmachine/validation/virtualmachine_validator.go

```diff
@@ -1951,3 +1953,121 @@ func (v validator) validateGroupName(
 
 	return allErrs
 }
+
+func (v validator) validateVMAffinity(
+	_ *pkgctx.WebhookRequestContext,
+	vm *vmopv1.VirtualMachine) field.ErrorList {
+
+	affinity := vm.Spec.Affinity
+	if affinity == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+
+	if vm.Spec.GroupName == "" {
+		allErrs = append(allErrs, field.Required(
+			field.NewPath("spec", "groupName"), "when setting affinity"))
+	}
+
+	path := field.NewPath("spec", "affinity")
+
+	if affinity.ZoneAffinity != nil {
+		allErrs = append(allErrs, field.Forbidden(path.Child("zoneAffinity"), "zone affinity is not allowed"))
+	}
+
+	if affinity.ZoneAntiAffinity != nil {
+		allErrs = append(allErrs, field.Forbidden(path.Child("zoneAntiAffinity"), "zone anti-affinity is not allowed"))
+	}
```

Comment:
Are these fields forbidden temporarily or always? Curious why we introduced them in spec if they cannot be set...